### PR TITLE
refactor: use DatabaseName in DatabaseRules

### PR DESCRIPTION
**Pull request by tustvold**
This turned into a bit of a yak shave as various tests were relying on being able to construct invalid DatabaseRules structures, but I think this is probably a good thing to fix

### CODE REVIEW

File: data_types/src/database_rules.rs [hunk 1]
```diff
@@ -589,12 +595,8 @@ pub struct PartitionTemplate {
     pub parts: Vec<TemplatePart>,
 }
 
-impl PartitionTemplate {
-    pub fn partition_key(
-        &self,
-        line: &ParsedLine<'_>,
-        default_time: &DateTime<Utc>,
-    ) -> Result<String> {
+impl Partitioner for PartitionTemplate {
```
Comment (tustvold) [inline data_types/src/database_rules.rs:598]: A number of tests were constructing a full DatabaseRules in order to just test the partitioning. As the partitioning is just determined by the PartitionTemplate it seemed sensible to just implement the trait for that

File: server/src/config.rs [hunk 2]
```diff
@@ -44,31 +44,19 @@ impl Config {
...
         let mut state = self.state.write().expect("mutex poisoned");
-        if state.reservations.contains(&name) || state.databases.contains_key(&name) {
+        if state.reservations.contains(&rules.name) || state.databases.contains_key(&rules.name) {
             return Err(Error::DatabaseAlreadyExists {
-                db_name: name.to_string(),
+                db_name: rules.name.to_string(),
             });
         }
 
-        let read_buffer = ReadBufferDb::new();
```
Comment (tustvold) [inline server/src/config.rs:57]: Defer creation of the Db until commit - this avoids some rwlock shenanigans

File: server/src/config.rs [hunk 3]
```diff
@@ -220,20 +217,25 @@ impl Drop for DatabaseState {
 /// persisted.
 #[derive(Debug)]
 pub(crate) struct CreateDatabaseHandle<'a> {
-    pub db: Arc<Db>,
-    pub name: DatabaseName<'static>,
+    rules: Option<DatabaseRules>,
```
Comment (tustvold) [inline server/src/config.rs:220]: Partial moves aren't supported on structures that implement Drop - so use Option to allow taking DatabaseRules out

File: server/src/lib.rs [hunk 4]
```diff
@@ -227,20 +227,13 @@ impl<M: ConnectionManager> Server<M> {
     /// Tells the server the set of rules for a database.
     pub async fn create_database(
         &self,
-        db_name: impl Into<String>,
-        mut rules: DatabaseRules,
+        rules: DatabaseRules,
```
Comment (tustvold) [inline server/src/lib.rs:230]: :tada:

File: server/src/lib.rs [hunk 5]
```diff
@@ -800,39 +793,13 @@ mod tests {
         Ok(())
     }
 
-    #[tokio::test]
```
Comment (tustvold) [inline server/src/lib.rs:803]: This test no longer makes sense - it isn't possible to construct a DatabaseRules with an invalid name

**Review by domodwyer — APPROVED** [review]

File: server/src/config.rs (hunk 3 above)
Comment (alamb) [reply server/src/config.rs:217]: that might be valuable to add as a comment in the code

**Review by alamb — APPROVED** [review]
looks like a nice improvement to me

**Comment by NGA-TRAN** [comment]
I will use this dyname in the paquert path today

**Pull request merged**